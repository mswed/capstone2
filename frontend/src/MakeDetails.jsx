const MakeDetails = () => {
  return <div>MakeDetails</div>;
};

export default MakeDetails;
