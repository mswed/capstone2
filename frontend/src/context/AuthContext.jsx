import React, { useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import GrumpyApi from '../services/api';

// We create the context
const AuthContext = React.createContext();

/**
 * This is a logic only component that wraps our entire app and sets the
 * authentication context and its functions
 *
 * @param {Array}  children - Components to wrap
 * @returns {AuthContext.Provider} - with auth functions for easy access
 */

const AuthProvider = ({ children }) => {
  // We grab stuff from local storage if available
  const storedToken = localStorage.getItem('grumpy-token');
  const storedUser = localStorage.getItem('grumpy-user');

  // We set the token right away so we don't run in to authorization issues
  GrumpyApi.token = storedToken || null;

  // Initialize state
  const [token, setToken] = useState(storedToken);
  const [currentUser, setCurrentuser] = useState(storedUser);

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    if (token && token.trim !== '') {
      // We only set the token if it's not empty

      GrumpyApi.token = token;
      localStorage.setItem('grumpy-token', token || '');
      localStorage.setItem('grumpy-user', currentUser || '');
    } else {
      GrumpyApi.token = null;
      localStorage.removeItem('grumpy-token');
      localStorage.removeItem('grumpy-user');
    }
  }, [token]);

  // TODO:
  // // We also want to get the user favorites here
  // useEffect(() => {
  //   if (!token) return;
  //   async function getUserData() {
  //     const userData = await GrumpuyApi.getUser(currentUser);
  //     setFavorites(new Set(userData.favorites));
  //   }
  //   getUserData();
  // }, [token]);

  const login = async (username, password) => {
    try {
      const authToken = await GrumpyApi.login(username, password);
      setToken(authToken);
      const decoded = jwtDecode(authToken);
      setCurrentuser(decoded.username);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await GrumpyApi.register(username, password, firstName, lastName, email);
      logout();
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = () => {
    // TODO: Do I need to logout of the backend here too?
    setToken('');
  };

  // TODO: We want to add favorites here
  // const addFavorite = async (jobId) => {
  //   const res = await GrumpuyApi.favoriteFormat(currentUser, jobId);
  //   if (res) {
  //     setFavorites((currentFavorites) => currentFavorites.add(res.favorites));
  //   }
  // };
  //

  // TODO: get items that have already been favorited
  // const alreadyFavorited = (formatId) => {
  //   return favFormats.has(formatId);
  // };

  // Create the value object
  const value = {
    token,
    currentUser,
    login,
    register,
    logout,
  };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
