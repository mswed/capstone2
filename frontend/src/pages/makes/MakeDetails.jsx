import { useState, useEffect, useContext } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { Container, Card, Row, Col } from 'react-bootstrap';
import GrumpyApi from '../../services/api';
import Loading from '../../components/ui/Loading';
import CameraList from '../../features/cameras/components/CameraList';
import ActionBar from '../../components/ui/ActionBar.jsx';
import ModalWindow from '../../components/ui/ModalWindow.jsx';
import MakeForm from '../../components/forms/MakeForm.jsx';
import CameraForm from '../../components/forms/CameraForm.jsx';
import ConfirmDialog from '../../components/ui/ConfirmDialog.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';
import { MessageContext } from '../../context/MessageContext.jsx';

const MakeDetails = () => {
  const navigate = useNavigate();
  const { showMessage } = useContext(MessageContext);

  // Setup state
  const { makeId } = useParams();
  const [makeData, setMakeData] = useState(null);
  const [showEditModal, setShowEditModal] = useState(false);
  const [showConfirmDelete, setConfirmDelete] = useState(false);
  const [showNewCameraModal, setShowNewCameraModal] = useState(false);
  const [isLoading, setIsLoading] = useState(true);

  const { token } = useContext(AuthContext);

  const actionButtons = [
    {
      text: 'Delete Make',
      variant: 'outline-danger',
      onClick: () => setConfirmDelete(true),
    },
    {
      text: 'Edit Make',
      variant: 'outline-warning',
      onClick: () => setShowEditModal(true),
    },
    {
      text: 'Add Camera',
      variant: 'outline-success',
      onClick: () => setShowNewCameraModal(true),
    },
  ];

  const handleEditMake = async (updatedMake) => {
    try {
      // Makes take an image as one of their fields so we have to
      // use FormData instead of json
      const formData = new FormData();
      formData.append('name', updatedMake.name);
      formData.append('website', updatedMake.website);
      if (updatedMake.logo) {
        formData.append('logo', updatedMake.logo);
      }

      const updatedData = await GrumpyApi.updateMake(makeId, formData);
      // Refresh the project
      setMakeData(updatedData);

      // Close the modal
      setShowEditModal(false);

      showMessage('Updated make', 'success');
    } catch (error) {
      showMessage('Failed to update make', 'danger');
      console.error('Failed to update make:', error);
    }
  };

  const handleDeleteMake = async () => {
    const response = await GrumpyApi.deleteMake(makeId);
    if (response.success) {
      navigate('/makes');
      showMessage('Deleted make', 'success');
    } else {
      showMessage('Failed to delete make', 'danger');
    }
  };

  const handleNewCamera = async (newCamera) => {
    try {
      // Cameras also take an image as one of their fields so we have to
      // use FormData instead of json
      const formData = new FormData();
      formData.append('make', makeId);
      formData.append('model', newCamera.model);
      formData.append('sensor_type', newCamera.sensorType);
      formData.append('max_filmback_width', newCamera.maxFilmbackWidth);
      formData.append('max_filmback_height', newCamera.maxFilmbackHeight);
      formData.append('max_image_width', newCamera.maxImageWidth);
      formData.append('max_image_height', newCamera.maxImageHeight);
      formData.append('min_frame_rate', newCamera.minFrameRate);
      formData.append('max_frame_rate', newCamera.maxFrameRate);
      formData.append('notes', newCamera.notes);
      formData.append('discontinued', newCamera.discontinued);
      if (newCamera.image) {
        formData.append('image', newCamera.image);
      }

      await GrumpyApi.addCamera(formData);

      // Refresh the make
      const response = await GrumpyApi.getMakeDetails(makeId);
      setMakeData(response);

      // Close the modal
      setShowNewCameraModal(false);
      showMessage('Added new camera', 'success');
    } catch (error) {
      showMessage('Failed to add new camera', 'success');
      console.error('Failed to add new camera:', error);
    }
  };

  useEffect(() => {
    const getMakeDetails = async () => {
      try {
        const response = await GrumpyApi.getMakeDetails(makeId);
        setMakeData(response);
      } catch (error) {
        console.error('Error fetching make details', error);
      } finally {
        setIsLoading(false);
      }
    };
    getMakeDetails();
  }, [makeId]);

  if (isLoading) {
    return <Loading />;
  }
  return (
    <Container>
      <Card className="shadow-lg rounded-0 rounded-bottom">
        <Row className="g-0">
          <Col md={4}>
            <Card.Img src={makeData.logo} alt={`Logo for ${makeData.name}`} className="h-100 p-3" style={{ objectFit: 'cover' }} />
          </Col>
          <Col md={8}>
            <Card.Body className="text-start">
              <Card.Title>{makeData.name}</Card.Title>
              <Card.Text>{makeData.website}</Card.Text>
            </Card.Body>
          </Col>
        </Row>
      </Card>
      {token && <ActionBar buttons={actionButtons} className="mt-3" />}
      <ConfirmDialog
        show={showConfirmDelete}
        title={`Delete make "${makeData.name}"`}
        message={'Are you sure?'}
        confirmText="Delete"
        onConfirm={handleDeleteMake}
        onCancel={() => setConfirmDelete(false)}
      />
      <ModalWindow show={showEditModal} onHide={() => setShowEditModal(false)} title={`Edit ${makeData.name}`} form={<MakeForm onSubmit={handleEditMake} makeData={makeData} />} />
      <ModalWindow
        show={showNewCameraModal}
        onHide={() => setShowNewCameraModal(false)}
        title={`New Camera By ${makeData.name}`}
        form={<CameraForm onSubmit={handleNewCamera} makeData={makeData} />}
      />
      {makeData.cameras.length > 0 && (
        <Row className="mt-3">
          <CameraList cameras={makeData.cameras} />
        </Row>
      )}
    </Container>
  );
};

export default MakeDetails;
