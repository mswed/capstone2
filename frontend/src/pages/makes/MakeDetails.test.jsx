import { vi, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import MakeDetails from './MakeDetails.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';
import { MessagesProvider } from '../../context/MessageContext.jsx';
import GrumpyApi from '../../services/api.js';

// Mock API calls
vi.mock('../../services/api', () => ({
  default: {
    getMakeDetails: vi.fn(() =>
      Promise.resolve({
        id: 1,
        name: 'ARRI',
        website: 'https://www.arri.com/en',
        logo: 'arri_logo.png',
        camerasCount: 2,
        cameras: [
          {
            id: 1,
            makeId: 1,
            makeName: 'Arri',
            model: 'Alexa 35',
            sensorType: 'Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array',
            sensorSize: 'Super 35',
            maxFilmbackWidth: 27.99,
            maxFilmbackHeight: 19.22,
            maxImageWidth: 4608,
            maxImageHeight: 3164,
            minFrameRate: 0.75,
            maxFrameRate: 120,
            image: 'arri_alexa_35.png',
          },
          {
            id: 2,
            makeId: 1,
            makeName: 'Arri',
            model: 'Alexa Mini LF',
            sensorType: 'Large Format ARRI ALEV III (A2X) CMOS sensor with Bayer pattern color filter array',
            sensorSize: 'Large Format',
            maxFilmbackWidth: 36.7,
            maxFilmbackHeight: 25.54,
            maxImageWith: 4448,
            maxImageHeight: 3096,
            minFrameRate: 0.75,
            maxFrameRate: 90,
            image: 'arri_alexa_mini.png',
          },
        ],
      })
    ),
    addCamera: vi.fn(() =>
      Promise.resolve({
        success: 'Created camera 7',
        cameraId: 7,
      })
    ),
    updateMake: vi.fn(() =>
      Promise.resolve({
        success: 'Partialy updated make',
        make: {
          id: 1,
          name: 'ARRI',
          website: 'https://www.arri.com/en',
          logo: 'arri_logo.png',
          camerasCount: 2,
          cameras: [
            {
              id: 1,
              makeId: 1,
              makeName: 'Arri',
              model: 'Alexa 35',
              sensorType: 'Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array',
              sensorSize: 'Super 35',
              maxFilmbackWidth: 27.99,
              maxFilmbackHeight: 19.22,
              maxImageWidth: 4608,
              maxImageHeight: 3164,
              minFrameRate: 0.75,
              maxFrameRate: 120,
              image: 'arri_alexa_35.png',
            },
            {
              id: 2,
              makeId: 1,
              makeName: 'Arri',
              model: 'Alexa Mini LF',
              sensorType: 'Large Format ARRI ALEV III (A2X) CMOS sensor with Bayer pattern color filter array',
              sensorSize: 'Large Format',
              maxFilmbackWidth: 36.7,
              maxFilmbackHeight: 25.54,
              maxImageWith: 4448,
              maxImageHeight: 3096,
              minFrameRate: 0.75,
              maxFrameRate: 90,
              image: 'arri_alexa_mini.png',
            },
          ],
        },
      })
    ),
    deleteMake: vi.fn(() =>
      Promise.resolve({
        success: 'Make deleted',
      })
    ),
  },
}));

describe('MakeDetails component', () => {
  /**
   * Render the componenet with authentication context
   *
   * @returns {Component} - Makes component wrapped in MemoryRouter and AuthContext
   */

  const renderWithAuth = (authValue = { token: 'fake-token', currentUser: 'testUser' }) => {
    return render(
      <MemoryRouter initialEntries={['/makes/1']}>
        <MessagesProvider>
          <AuthContext.Provider value={authValue}>
            <Routes>
              <Route path="/makes/:makeId" element={<MakeDetails />} />
            </Routes>
          </AuthContext.Provider>
        </MessagesProvider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    renderWithAuth();
    await screen.findByText('ARRI');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderWithAuth();

    await screen.findByText('ARRI');
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct info', async () => {
    renderWithAuth();

    await waitFor(() => {
      expect(screen.getByText('ARRI')).toBeInTheDocument();
      expect(screen.getByText('https://www.arri.com/en')).toBeInTheDocument();
      expect(screen.getByText('Alexa 35')).toBeInTheDocument();
      expect(screen.getByText('Alexa Mini LF')).toBeInTheDocument();
    });
  });

  it('shows action bar for admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'admminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');
    expect(screen.getByText('Add Camera')).toBeInTheDocument();
  });

  it('hides action bar for regular users', async () => {
    renderWithAuth({ token: 'fake-token', currentUser: 'admminUser' });
    await screen.findByText('ARRI');
    expect(screen.queryByText('Add Camera')).not.toBeInTheDocument();
  });

  it('can add a new camera when admin', async () => {
    // TODO: Might need to mock CameraForm component to speed up this test
    // Currently ~578ms due to filling out 8 form fields

    renderWithAuth({
      token: 'fake-token',
      currentUser: 'adminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');

    // Open modal
    fireEvent.click(screen.getByText('Add Camera'));
    expect(screen.getByRole('textbox', { name: /Model/i })).toBeInTheDocument();
    expect(screen.getByRole('textbox', { name: /Sensor Type/i })).toBeInTheDocument();

    // Fill out the required fields
    const modelInput = screen.getByLabelText(/Model/i);
    const sensorSizeInput = screen.getByLabelText(/Sensor Type/i);
    const maxFilmbackWidthInput = screen.getByLabelText(/Max Filmback Width/i);
    const maxFilmbackHeightInput = screen.getByLabelText(/Max Filmback Height/i);
    const maxImageWidthInput = screen.getByLabelText(/Max Image Width/i);
    const maxImageHeightInput = screen.getByLabelText(/Max Image Height/i);
    const maxFrameRateInput = screen.getByLabelText(/Max Frame Rate/i);
    const minFrameRateInput = screen.getByLabelText(/Min Frame Rate/i);

    fireEvent.change(modelInput, { target: { value: 'Alexa Amira' } });
    fireEvent.change(sensorSizeInput, {
      target: {
        value: 'Super 35 format ARRI ALEV III CMOS sensor with Bayer pattern color filter array',
      },
    });
    fireEvent.change(maxFilmbackWidthInput, {
      target: {
        value: 26.4,
      },
    });
    fireEvent.change(maxFilmbackHeightInput, {
      target: {
        value: 14.85,
      },
    });
    fireEvent.change(maxImageWidthInput, {
      target: { value: 3200 },
    });
    fireEvent.change(maxImageHeightInput, {
      target: {
        value: 1800,
      },
    });
    fireEvent.change(minFrameRateInput, {
      target: {
        value: 0.75,
      },
    });
    fireEvent.change(maxFrameRateInput, {
      target: {
        value: 200,
      },
    });

    // Submit the form
    fireEvent.click(screen.getByText('Add'));

    // Modal should close after successful submission
    await waitFor(() => {
      expect(screen.queryByText('New Camera By Arri')).not.toBeInTheDocument();
    });
  });

  it('can edit the make when admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'adminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');

    // Open modal
    fireEvent.click(screen.getByText('Edit Make'));
    expect(screen.getByText('Website')).toBeInTheDocument();

    // Fill out the required fields
    const nameInput = screen.getByLabelText(/name/i);

    fireEvent.change(nameInput, { target: { value: 'Updated Make' } });

    // Submit the form
    fireEvent.click(screen.getByText('Update'));

    // Modal should close after successful submission
    await waitFor(() => {
      expect(screen.queryByText('Website')).not.toBeInTheDocument();
    });
  });

  it('can delete the make when admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'adminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');

    // Open modal
    fireEvent.click(screen.getByText('Delete Make'));
    expect(screen.getByText('Delete make "ARRI"')).toBeInTheDocument();

    // Click the button
    fireEvent.click(screen.getByText('Delete'));

    // Modal should close after successful submission
    await waitFor(() => {
      expect(GrumpyApi.deleteMake).toHaveBeenCalledWith('1');
    });
  });
});
