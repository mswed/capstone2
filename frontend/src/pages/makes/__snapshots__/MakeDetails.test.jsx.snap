// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MakeDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="shadow-lg rounded-0 rounded-bottom card"
    >
      <div
        class="g-0 row"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="Logo for ARRI"
            class="card-img h-100 p-3"
            src="arri_logo.png"
            style="object-fit: cover;"
          />
        </div>
        <div
          class="col-md-8"
        >
          <div
            class="text-start card-body"
          >
            <div
              class="card-title h5"
            >
              ARRI
            </div>
            <p
              class="card-text"
            >
              https://www.arri.com/en
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mt-3 row"
    >
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-hover"
        >
          <thead>
            <tr>
              <th>
                Image
              </th>
              <th>
                Model
              </th>
              <th>
                Sensor Type
              </th>
              <th>
                Max Filmback Size
              </th>
              <th>
                Max Resolution
              </th>
              <th>
                Action
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <img
                  alt="Image of Alexa 35"
                  src="arri_alexa_35.png"
                  style="width: 60px;"
                />
              </td>
              <td>
                Alexa 35
              </td>
              <td />
              <td>
                mm x mm
              </td>
              <td>
                 x 
              </td>
              <td>
                <a
                  class="btn btn-outline-primary btn-sm"
                  data-discover="true"
                  href="/cameras/1"
                >
                  View
                </a>
              </td>
            </tr>
            <tr>
              <td>
                <img
                  alt="Image of Alexa Mini LF"
                  src="arri_alexa_mini.png"
                  style="width: 60px;"
                />
              </td>
              <td>
                Alexa Mini LF
              </td>
              <td />
              <td>
                mm x mm
              </td>
              <td>
                 x 
              </td>
              <td>
                <a
                  class="btn btn-outline-primary btn-sm"
                  data-discover="true"
                  href="/cameras/2"
                >
                  View
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
